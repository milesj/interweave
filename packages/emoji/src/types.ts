import { Emoji, Emoticon, Hexcode, Locale, Shortcode, Unicode } from 'emojibase';
import { MatchParams } from 'interweave';

export interface CanonicalEmoji extends Emoji {
	canonicalShortcodes: Shortcode[];
	primaryShortcode: Shortcode;
	skins?: CanonicalEmoji[];
	unicode: Unicode;
}

export interface PathConfig {
	enlarged: boolean;
	largeSize: Size;
	size: Size;
	smallSize: Size;
}

export type Path = string | ((hexcode: Hexcode, config: PathConfig) => string);

export type Size = number | string;

export interface Source {
	compact: boolean;
	error?: Error;
	locale: Locale;
	version: string;
}

export interface EmojiProps extends EmojiConfig {
	/** Emoticon to reference emoji from. */
	emoticon?: Emoticon;
	/** Enlarge emoji increasing it's size. */
	enlarge?: boolean;
	/** Hexcode to reference emoji from. */
	hexcode?: Hexcode;
	/** Path to an SVG/PNG. Accepts a string or a callback that is passed the hexcode. */
	path?: Path;
	/** Shortcode to reference emoji from. */
	shortcode?: Shortcode;
	/** Emoji datasource metadata. */
	source: Source;
	/** Unicode character to reference emoji from. */
	unicode?: Unicode;
}

export interface EmojiMatch extends MatchParams {
	emoticon?: string;
	hexcode?: string;
	shortcode?: string;
	unicode?: string;
}

export interface EmojiConfig {
	/** Size of the emoji when it's enlarged. */
	largeSize?: Size;
	/** Render literal unicode character instead of an SVG/PNG. */
	renderUnicode?: boolean;
	/** Size of the emoji. Defaults to 1em. */
	size?: Size;
}

export interface UseEmojiDataOptions {
	/** Avoid fetching emoji data. Assumes data has already been fetched. */
	avoidFetch?: boolean;
	/** Load compact emoji dataset instead of full dataset. */
	compact?: boolean;
	/** Locale to load emoji labels in. Defaults to `en`. */
	locale?: Locale;
	/** List of shortcode presets to load and join with the dataset. */
	shortcodes?: string[];
	/** Throw errors that occurred during a fetch. Defaults to `true`. */
	throwErrors?: boolean;
	/** Emojibase dataset version to load. Defaults to `latest`. */
	version?: string;
}

declare module 'interweave' {
	export interface InterweaveProps {
		emojiEnlargeThreshold?: number;
		emojiSource: Source;
	}
}
