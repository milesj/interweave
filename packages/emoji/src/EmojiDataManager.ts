import {
	Emoji,
	EMOTICON_OPTIONS,
	fromCodepointToUnicode,
	fromHexcodeToCodepoint,
	generateEmoticonPermutations,
	GroupKey,
	Hexcode,
	Locale,
	MessagesDataset,
	SkinToneKey,
	SubgroupKey,
	TEXT,
} from 'emojibase';
import { CanonicalEmoji } from './types';

const instances = new Map<string, EmojiDataManager>();

export function resetInstances() {
	if (__DEV__) {
		instances.clear();
	}
}

export class EmojiDataManager {
	EMOJIS: Record<string, CanonicalEmoji> = {};

	EMOTICON_TO_HEXCODE: Record<string, Hexcode> = {};

	SHORTCODE_TO_HEXCODE: Record<string, Hexcode> = {};

	UNICODE_TO_HEXCODE: Record<string, Hexcode> = {};

	GROUPS_BY_KEY: Partial<Record<GroupKey, string>> = {};

	SKIN_TONES_BY_KEY: Partial<Record<SkinToneKey, string>> = {};

	SUBGROUPS_BY_KEY: Partial<Record<SubgroupKey, string>> = {};

	data: CanonicalEmoji[] = [];

	flatData: CanonicalEmoji[] = [];

	locale: Locale = 'en';

	version: string = '0.0.0';

	constructor(locale: Locale, version: string) {
		this.locale = locale;
		this.version = version;
	}

	/**
	 * Return or create a singleton instance per locale.
	 */
	static getInstance(locale: Locale, version: string): EmojiDataManager {
		const key = `${locale}:${version}`;

		if (!instances.has(key)) {
			instances.set(key, new EmojiDataManager(locale, version));
		}

		return instances.get(key)!;
	}

	/**
	 * Return dataset as a list.
	 */
	getData(): CanonicalEmoji[] {
		return this.data;
	}

	/**
	 * Return dataset as a flattened list.
	 */
	getFlatData(): CanonicalEmoji[] {
		return this.flatData;
	}

	/**
	 * Package the emoji object with additional data,
	 * while also extracting and partitioning relevant information.
	 */
	packageEmoji(baseEmoji: Emoji): CanonicalEmoji {
		const { emoticon, hexcode, shortcodes = [] } = baseEmoji;
		const emoji: CanonicalEmoji = {
			...baseEmoji,
			canonicalShortcodes: [],
			primaryShortcode: '',
			skins: [],
			unicode: '',
		};

		// Make our lives easier
		if (!emoji.unicode) {
			emoji.unicode = emoji.text && emoji.type === TEXT ? emoji.text : emoji.emoji;
		}

		// Canonicalize the shortcodes for easy reuse
		emoji.canonicalShortcodes = shortcodes.map((code) => `:${code}:`);
		emoji.primaryShortcode = emoji.canonicalShortcodes[0];

		// Support all shortcodes
		emoji.canonicalShortcodes.forEach((shortcode) => {
			this.SHORTCODE_TO_HEXCODE[shortcode] = hexcode;
		});

		// Support all emoticons
		if (emoticon) {
			const emoticons = Array.isArray(emoticon) ? emoticon : [emoticon];

			emoticons.forEach((emo) => {
				generateEmoticonPermutations(emo, EMOTICON_OPTIONS[emo]).forEach((e) => {
					this.EMOTICON_TO_HEXCODE[e] = hexcode;
				});
			});
		}

		// Support all presentations (even no variation selectors)
		this.UNICODE_TO_HEXCODE[fromCodepointToUnicode(fromHexcodeToCodepoint(hexcode))] = hexcode;

		if (emoji.emoji) {
			this.UNICODE_TO_HEXCODE[emoji.emoji] = hexcode;
		}

		if (emoji.text) {
			this.UNICODE_TO_HEXCODE[emoji.text] = hexcode;
		}

		// Map each emoji
		this.EMOJIS[hexcode] = emoji;

		// Apply same logic to all variations
		if (baseEmoji.skins) {
			emoji.skins = baseEmoji.skins.map((skinEmoji) => this.packageEmoji(skinEmoji));
		}

		return emoji;
	}

	/**
	 * Parse and generate emoji datasets.
	 */
	parseEmojiData(data: Emoji[]): CanonicalEmoji[] {
		data.forEach((emoji) => {
			const packagedEmoji = this.packageEmoji(emoji);

			this.data.push(packagedEmoji);
			this.flatData.push(packagedEmoji);

			// Flatten and package skins as well
			if (packagedEmoji.skins) {
				packagedEmoji.skins.forEach((skin) => {
					this.flatData.push(skin);
				});
			}
		});

		return this.data;
	}

	parseMessageData(data: MessagesDataset) {
		if (data.groups) {
			data.groups.forEach((group) => {
				this.GROUPS_BY_KEY[group.key] = group.message;
			});
		}

		if (data.subgroups) {
			data.subgroups.forEach((group) => {
				this.SUBGROUPS_BY_KEY[group.key] = group.message;
			});
		}

		if (data.skinTones) {
			data.skinTones.forEach((skinTone) => {
				this.SKIN_TONES_BY_KEY[skinTone.key] = skinTone.message;
			});
		}
	}
}
