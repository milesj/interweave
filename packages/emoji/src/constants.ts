// v13.1 isnt available on platforms yet
export const MAX_EMOJI_VERSION = 13;

// v5 is for emoji version v13 (above)
export const LATEST_DATASET_VERSION = '5.1.0';
