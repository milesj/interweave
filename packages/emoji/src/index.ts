/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './constants';
export * from './Emoji';
export * from './EmojiDataManager';
export * from './matchers';
export * from './types';
export * from './useEmojiData';
