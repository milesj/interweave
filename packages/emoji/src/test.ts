import { Locale } from 'emojibase';
import { loadEmojiData, loadMeta, loadShortcodes } from 'emojibase-test-utils';
// eslint-disable-next-line unicorn/import-index
import { CanonicalEmoji, EmojiDataManager } from './index';

export function mockEmojiData(locale: Locale = 'en'): CanonicalEmoji[] {
	const instance = EmojiDataManager.getInstance(locale);

	if (instance.data.length === 0) {
		instance.parseEmojiData(loadEmojiData([loadShortcodes()]));
		instance.parseMessageData(loadMeta());
	}

	return instance.getData();
}
