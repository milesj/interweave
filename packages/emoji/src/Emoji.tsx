/* eslint-disable complexity */

import React from 'react';
import { EmojiDataManager } from './EmojiDataManager';
import { EmojiProps, Size } from './types';

export function Emoji({
	emojiLargeSize = '3em',
	emojiPath = '{{hexcode}}',
	emojiSize = '1em',
	emojiSource,
	emoticon,
	enlargeEmoji = false,
	hexcode,
	renderUnicode = false,
	shortcode,
	unicode,
}: EmojiProps) {
	const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);

	if (__DEV__ && !emoticon && !shortcode && !unicode && !hexcode) {
		throw new Error(
			'Emoji component requires a `unicode` character, `emoticon`, `hexcode`, or a `shortcode`.',
		);
	}

	// Retrieve applicable unicode character
	let hex = hexcode;

	if (!hex && shortcode) {
		hex = data.SHORTCODE_TO_HEXCODE[shortcode];
	}

	if (!hex && emoticon) {
		hex = data.EMOTICON_TO_HEXCODE[emoticon];
	}

	if (!hex && unicode) {
		hex = data.UNICODE_TO_HEXCODE[unicode];
	}

	// Return the invalid value instead of erroring
	if (!hex || !data.EMOJIS[hex]) {
		return <span>{unicode ?? emoticon ?? shortcode ?? hex}</span>;
	}

	const emoji = data.EMOJIS[hex];

	if (renderUnicode) {
		return <span>{emoji.unicode}</span>;
	}

	// eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
	const styles: Record<string, Size | string> = {
		display: 'inline-block',
		verticalAlign: 'middle',
	};

	// Handle large styles
	if (enlargeEmoji && emojiLargeSize) {
		styles.width = emojiLargeSize;
		styles.height = emojiLargeSize;

		// Only apply styles if a size is defined
	} else if (emojiSize) {
		styles.width = emojiSize;
		styles.height = emojiSize;
	}

	// Determine the path
	let path = emojiPath || '{{hexcode}}';

	path =
		typeof path === 'function'
			? path(emoji.hexcode, {
					enlarged: enlargeEmoji,
					largeSize: emojiLargeSize,
					size: enlargeEmoji ? emojiLargeSize : emojiSize,
					smallSize: emojiSize,
			  })
			: path.replace('{{hexcode}}', emoji.hexcode);

	// http://git.emojione.com/demos/latest/sprites-png.html
	// http://git.emojione.com/demos/latest/sprites-svg.html
	// https://css-tricks.com/using-svg/
	return (
		<img
			alt={emoji.unicode}
			aria-label={emoji.label}
			data-emoticon={emoji.emoticon}
			data-hexcode={emoji.hexcode}
			data-shortcodes={emoji.canonical_shortcodes.join(', ')}
			data-unicode={emoji.unicode}
			src={path}
			style={styles}
			title={emoji.label}
		/>
	);
}
