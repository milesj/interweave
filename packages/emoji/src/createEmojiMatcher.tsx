import React from 'react';
import {
	createMatcher,
	InterweaveProps,
	MatcherFactory,
	MatcherFactoryData,
	Node,
	OnMatch,
} from 'interweave';
import { DEFAULT_CONFIG } from './constants';
import { Emoji } from './Emoji';
import { EmojiConfig, EmojiMatch } from './types';

function factory({
	config,
	params,
	props: { emojiSource },
}: MatcherFactoryData<EmojiMatch, InterweaveProps, EmojiConfig>) {
	return <Emoji {...config} {...params} source={emojiSource} />;
}

function onBeforeParse(content: string, { emojiSource }: InterweaveProps): string {
	if (__DEV__ && !emojiSource) {
		throw new Error(
			'Missing emoji source data. Have you loaded with the `useEmojiData` hook and passed the `emojiSource` prop?',
		);
	}

	return content;
}

function onAfterParse(node: Node, { emojiEnlargeThreshold = 1 }: InterweaveProps): Node {
	const content = React.Children.toArray(node);

	if (content.length === 0) {
		return content;
	}

	let valid = false;
	let count = 0;

	// Use a for-loop, as it's much cleaner than some()
	for (const item of content) {
		if (typeof item === 'string') {
			// Allow whitespace but disallow strings
			if (!item.match(/^\s+$/)) {
				valid = false;
				break;
			}
		} else if (React.isValidElement(item)) {
			// Only count towards emojis
			if (item && item.type === Emoji) {
				count += 1;
				valid = true;

				if (count > emojiEnlargeThreshold) {
					valid = false;
					break;
				}

				// Abort early for non-emoji components
			} else {
				valid = false;
				break;
			}
		} else {
			valid = false;
			break;
		}
	}

	if (!valid) {
		return content;
	}

	return content.map((item) => {
		if (!React.isValidElement(item)) {
			return item;
		}

		return React.cloneElement(item, {
			...(item.props as {}),
			enlarge: true,
		});
	});
}

export function createEmojiMatcher(
	pattern: RegExp,
	onMatch: OnMatch<EmojiMatch, InterweaveProps>,
	customFactory: MatcherFactory<EmojiMatch, InterweaveProps, EmojiConfig> = factory,
) {
	return createMatcher<EmojiMatch, InterweaveProps, EmojiConfig>(pattern, customFactory, {
		config: DEFAULT_CONFIG,
		greedy: true,
		onAfterParse,
		onBeforeParse,
		onMatch,
		tagName: 'img',
		void: true,
	});
}
