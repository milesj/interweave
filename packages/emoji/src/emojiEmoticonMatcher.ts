import EMOTICON_REGEX from 'emojibase-regex/emoticon';
import EmojiDataManager from './EmojiDataManager';
import createEmojiMatcher from './createEmojiMatcher';

const EMOTICON_BOUNDARY_REGEX = new RegExp(
  // eslint-disable-next-line no-useless-escape
  `(^|\\\b|\\\s)(${EMOTICON_REGEX.source})(?=\\\s|\\\b|$)`,
);

export default createEmojiMatcher(EMOTICON_BOUNDARY_REGEX, (result, { emojiSource }) => {
  const data = EmojiDataManager.getInstance(emojiSource.locale);
  const emoticon = result.matches[0].trim();

  if (!emoticon || !data.EMOTICON_TO_HEXCODE[emoticon]) {
    return null;
  }

  // Remove padding
  // TODO still needed?
  result.match = String(emoticon);

  return {
    emoticon,
    hexcode: data.EMOTICON_TO_HEXCODE[emoticon],
  };
});
