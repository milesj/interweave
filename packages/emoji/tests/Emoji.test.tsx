import React from 'react';
import { SOURCE_PROP, VALID_EMOJIS } from 'interweave/test';
import { render } from 'rut-dom';
import { Emoji } from '../src/Emoji';
import { EmojiProps } from '../src/types';

describe('Emoji', () => {
	const [[hexcode, unicode, shortcode, emoticon]] = VALID_EMOJIS;

	it('errors if no emoticon, shortcode or unicode', () => {
		expect(() => render<EmojiProps>(<Emoji source={SOURCE_PROP} />)).toThrowErrorMatchingSnapshot();
	});

	it('returns value for invalid hexcode', () => {
		const { root } = render<EmojiProps>(<Emoji hexcode="FA" source={SOURCE_PROP} />);

		expect(root.findOne('span')).toMatchSnapshot();
	});

	it('returns value for invalid emoticon', () => {
		const { root } = render<EmojiProps>(<Emoji emoticon="0P" source={SOURCE_PROP} />);

		expect(root.findOne('span')).toMatchSnapshot();
	});

	it('returns value for invalid shortcode', () => {
		const { root } = render<EmojiProps>(<Emoji shortcode="fake" source={SOURCE_PROP} />);

		expect(root.findOne('span')).toMatchSnapshot();
	});

	it('returns empty for invalid unicode', () => {
		const { root } = render<EmojiProps>(<Emoji source={SOURCE_PROP} unicode="fake" />);

		expect(root.findOne('span')).toMatchSnapshot();
	});

	it('renders with only the emoticon', () => {
		const { root } = render<EmojiProps>(<Emoji emoticon={emoticon} source={SOURCE_PROP} />);

		expect(root.findOne('img')).toMatchSnapshot();
	});

	it('renders with only the shortcode', () => {
		const { root } = render<EmojiProps>(<Emoji shortcode={shortcode} source={SOURCE_PROP} />);

		expect(root.findOne('img')).toMatchSnapshot();
	});

	it('renders with only the hexcode', () => {
		const { root } = render<EmojiProps>(<Emoji hexcode={hexcode} source={SOURCE_PROP} />);

		expect(root.findOne('img')).toMatchSnapshot();
	});

	it('renders with only the unicode', () => {
		const { root } = render<EmojiProps>(<Emoji source={SOURCE_PROP} unicode={unicode} />);

		expect(root.findOne('img')).toMatchSnapshot();
	});

	it('renders with both', () => {
		const { root } = render<EmojiProps>(
			<Emoji shortcode={shortcode} source={SOURCE_PROP} unicode={unicode} />,
		);

		expect(root.findOne('img')).toMatchSnapshot();
	});

	it('renders the unicode character', () => {
		const { root } = render<EmojiProps>(
			<Emoji renderUnicode source={SOURCE_PROP} unicode={unicode} />,
		);

		expect(root.findOne('span')).toMatchSnapshot();
	});

	it('can define the path', () => {
		const { root } = render<EmojiProps>(
			<Emoji
				path="http://foo.com/path/to/{{hexcode}}.svg"
				shortcode={shortcode}
				source={SOURCE_PROP}
				unicode={unicode}
			/>,
		);
		const img = root.findOne('img');

		expect(img).toHaveProp('alt', unicode);
		expect(img).toHaveProp('src', `http://foo.com/path/to/${hexcode}.svg`);
	});

	it('can define the path with a function', () => {
		const { root } = render<EmojiProps>(
			<Emoji
				path={(hex) => `http://foo.com/path/to/${hex.toLowerCase()}.svg`}
				shortcode={shortcode}
				source={SOURCE_PROP}
				unicode={unicode}
			/>,
		);
		const img = root.findOne('img');

		expect(img).toHaveProp('alt', unicode);
		expect(img).toHaveProp('src', `http://foo.com/path/to/${hexcode.toLowerCase()}.svg`);
	});

	it('path function receives enlarge and size', () => {
		const { root } = render<EmojiProps>(
			<Emoji
				enlarge
				largeSize={4}
				path={(hex, { size }) => `http://foo.com/path/to/${size}/${hex.toLowerCase()}.svg`}
				shortcode={shortcode}
				size={2}
				source={SOURCE_PROP}
				unicode={unicode}
			/>,
		);

		expect(root.findOne('img')).toHaveProp(
			'src',
			`http://foo.com/path/to/4/${hexcode.toLowerCase()}.svg`,
		);
	});

	it('sets styles when size is defined', () => {
		const { root } = render<EmojiProps>(
			<Emoji shortcode={shortcode} size={1} source={SOURCE_PROP} unicode={unicode} />,
		);

		expect(root.findOne('img')).toHaveProp('style', {
			display: 'inline-block',
			verticalAlign: 'middle',
			width: 1,
			height: 1,
		});
	});

	it('can customize large size', () => {
		const { root } = render<EmojiProps>(
			<Emoji
				enlarge
				largeSize={5}
				shortcode={shortcode}
				size={2}
				source={SOURCE_PROP}
				unicode={unicode}
			/>,
		);

		expect(root.findOne('img')).toHaveProp('style', {
			display: 'inline-block',
			verticalAlign: 'middle',
			width: 5,
			height: 5,
		});
	});

	it('can use string sizes', () => {
		const { root } = render<EmojiProps>(
			<Emoji shortcode={shortcode} size="2em" source={SOURCE_PROP} unicode={unicode} />,
		);

		expect(root.findOne('img')).toHaveProp('style', {
			display: 'inline-block',
			verticalAlign: 'middle',
			width: '2em',
			height: '2em',
		});
	});
});
