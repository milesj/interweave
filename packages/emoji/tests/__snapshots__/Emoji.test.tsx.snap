// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoji errors if no emoticon, shortcode or unicode 1`] = `"Emoji component requires a \`unicode\` character, \`emoticon\`, \`hexcode\`, or a \`shortcode\`."`;

exports[`Emoji renders the unicode character 1`] = `<span>ðŸ˜¡</span>`;

exports[`Emoji renders with both 1`] = `
<img
  alt="ðŸ˜¡"
  aria-label="pouting face"
  data-emoticon=">:/"
  data-hexcode="1F621"
  data-shortcodes=":pout:, :pouting_face:, :rage:"
  data-unicode="ðŸ˜¡"
  src="1F621"
  style={{ display: "inline-block", verticalAlign: "middle", width: "1em", height: "1em" }}
  title="pouting face"
/>
`;

exports[`Emoji renders with only the emoticon 1`] = `
<img
  alt="ðŸ˜¡"
  aria-label="pouting face"
  data-emoticon=">:/"
  data-hexcode="1F621"
  data-shortcodes=":pout:, :pouting_face:, :rage:"
  data-unicode="ðŸ˜¡"
  src="1F621"
  style={{ display: "inline-block", verticalAlign: "middle", width: "1em", height: "1em" }}
  title="pouting face"
/>
`;

exports[`Emoji renders with only the hexcode 1`] = `
<img
  alt="ðŸ˜¡"
  aria-label="pouting face"
  data-emoticon=">:/"
  data-hexcode="1F621"
  data-shortcodes=":pout:, :pouting_face:, :rage:"
  data-unicode="ðŸ˜¡"
  src="1F621"
  style={{ display: "inline-block", verticalAlign: "middle", width: "1em", height: "1em" }}
  title="pouting face"
/>
`;

exports[`Emoji renders with only the shortcode 1`] = `
<img
  alt="ðŸ˜¡"
  aria-label="pouting face"
  data-emoticon=">:/"
  data-hexcode="1F621"
  data-shortcodes=":pout:, :pouting_face:, :rage:"
  data-unicode="ðŸ˜¡"
  src="1F621"
  style={{ display: "inline-block", verticalAlign: "middle", width: "1em", height: "1em" }}
  title="pouting face"
/>
`;

exports[`Emoji renders with only the unicode 1`] = `
<img
  alt="ðŸ˜¡"
  aria-label="pouting face"
  data-emoticon=">:/"
  data-hexcode="1F621"
  data-shortcodes=":pout:, :pouting_face:, :rage:"
  data-unicode="ðŸ˜¡"
  src="1F621"
  style={{ display: "inline-block", verticalAlign: "middle", width: "1em", height: "1em" }}
  title="pouting face"
/>
`;

exports[`Emoji returns empty for invalid unicode 1`] = `<span>fake</span>`;

exports[`Emoji returns value for invalid emoticon 1`] = `<span>0P</span>`;

exports[`Emoji returns value for invalid hexcode 1`] = `<span>FA</span>`;

exports[`Emoji returns value for invalid shortcode 1`] = `<span>fake</span>`;
